File: 804.unique-morse-code-words.cpp
/*
 * @lc app=leetcode id=804 lang=cpp
 *
 * [804] Unique Morse Code Words
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
static const vector<string> table_ = {
    ".-",   "-...", "-.-.", "-..",  ".",   "..-.", "--.",  "....", "..",
    ".---", "-.-",  ".-..", "--",   "-.",  "---",  ".--.", "--.-", ".-.",
    "...",  "-",    "..-",  "...-", ".--", "-..-", "-.--", "--.."};
class Solution {
 public:
  int uniqueMorseRepresentations(vector<string>& words) {
    std::set<string> morse_words;
    for (const auto& word : words) {
      morse_words.insert(morse(word));
    }
    return morse_words.size();
  }

 private:
  std::string morse(const std::string word) {
    std::string ret;
    for (const auto c : word) {
      ret += table_[c - 'a'];
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<string> words;
  words = {"gin", "zen", "gig", "msg"};
  ASSERT_EQ(s.uniqueMorseRepresentations(words), 2);
  words = {""};
  ASSERT_EQ(s.uniqueMorseRepresentations(words), 1);
  words = {"", ""};
  ASSERT_EQ(s.uniqueMorseRepresentations(words), 1);
  return 0;
}

File: 804.unique-morse-code-words.rs
/*
 * @lc app=leetcode id=804 lang=rust
 *
 * [804] Unique Morse Code Words
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
use std::collections::HashSet;
impl Solution {
    pub fn unique_morse_representations(words: Vec<String>) -> i32 {
        static MORSE: [&str; 26] = [
            ".-", "-...", "-.-.", "-..", ".", "..-.", "--.", "....", "..", ".---", "-.-", ".-..",
            "--", "-.", "---", ".--.", "--.-", ".-.", "...", "-", "..-", "...-", ".--", "-..-",
            "-.--", "--..",
        ];

        words
            .into_iter()
            .map(|word| {
                word.chars()
                    .map(|c| MORSE[c as usize - 'a' as usize])
                    .collect::<String>()
            })
            .collect::<HashSet<String>>()
            .len() as _
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(
        Solution::unique_morse_representations(vec![
            "gin".to_string(),
            "zen".to_string(),
            "gig".to_string(),
            "msg".to_string()
        ]),
        2
    );
}

#[test]
fn test2() {
    assert_eq!(
        Solution::unique_morse_representations(vec!["a".to_string()]),
        1
    );
}
